cosmos/scheduler/cmd/scheduler.go

```diff
@@ -0,0 +1,265 @@
+		log.Println("got an overlay update")
+		quit <- true
+	}
+}
+
+func (s *Scheduler) configureIngestionSecret() error {
+	secret, err := s.getIngestionSecret()
+	if err != nil {
+		return err
+	}
+	if secret != "" {
+		s.SecretName = secret
+		log.Println("ingestion credentials secret found")
+		return nil
+	}
+	log.Println("ingestion secret not found")
+	secret, err = s.createIngestionSecret()
+	if err != nil {
+		return err
+	}
+	s.SecretName = secret
+	log.Println("ingestion secret created successfully")
+	return nil
+}
+
+func (s *Scheduler) run(quit chan bool) error {
+	locationBson, err := s.getCosmosLocationBson()
+	if err != nil {
+		return err
+	}
+	if len(locationBson) == 0 {
+		return nil
+	}
+	collection := s.MongodbClient.Database("metadata").Collection("__metastore")
+	ctx := context.Background()
+	cur, err := collection.Watch(ctx, mongo.Pipeline{
+		{{"$match", bson.D{
+			{"$or", locationBson},
+			{"fullDocument.value.ingestion.status", "enabled"},
+		}}},
+		{{"$project", bson.D{
+			{"operationType", 1},
+			{"fullDocument.value.ownerDisplayName", 1},
+			{"fullDocument.value.name", 1},
+			{"fullDocument.value.locationConstraint", 1},
+			{"fullDocument.value.deleted", 1},
+		}}},
+	})
+	if err != nil {
+		log.Println(err)
+		return err
+	}
+	ch := make(chan BucketTransaction)
+	go func() {
+		defer cur.Close(ctx)
+		for {
+			if cur.Next(ctx) {
+				var elem BucketTransaction
+				if err := cur.Decode(&elem); err != nil {
+					log.Println(err)
+					return
+				}
+				ch <- elem
+			} else {
+				return
+			}
+			if err := cur.Err(); err != nil {
+				log.Println(err)
+				return
+			}
+		}
+	}()
+	for {
+		select {
+		case <-quit:
+			cur.Close(ctx)
+			return nil
+		case bucket := <-ch:
+			s.applyChanges(&bucket)
+		}
+	}
+}
+
+func (s *Scheduler) watchOverlayUpdates() chan interface{} {
+	ch := make(chan interface{})
+	collection := s.MongodbClient.Database("metadata").Collection("PENSIEVE")
+	ctx := context.Background()
+	cur, err := collection.Watch(ctx, mongo.Pipeline{})
+	if err != nil {
+		log.Println(err)
+		close(ch)
+		return nil
+	}
+	go func() {
+		defer cur.Close(ctx)
+		for {
+			if cur.Next(ctx) {
+				ch <- "overlay update"
+			}
+			if err := cur.Err(); err != nil {
+				log.Println(err)
+			}
+		}
+	}()
+	return ch
+}
+
+func (s *Scheduler) getIngestionSecret() (string, error) {
+	secrets, err := s.KubeClientset.CoreV1().
+		Secrets(s.Namespace).
+		List(metav1.ListOptions{LabelSelector: "ingestionCredentials=true"})
+	if err != nil {
+		return "", nil
+	}
+	if len(secrets.Items) > 0 {
+		return secrets.Items[0].Name, nil
+	}
+	return "", nil
+}
+
+func (s *Scheduler) createIngestionSecret() (string, error) {
+	accessKey, secretKey, err := s.Pensieve.GetServiceAccountCredentials("md-ingestion")
+	if err != nil {
+		return "", err
+	}
+	secret, err := s.KubeClientset.CoreV1().Secrets(s.Namespace).Create(&v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "cosmos-metadata-ingestion",
```

Comment:
Add a todo here as it would conflict if there are multiple Zenkos within one namespace.

Reply:
Done